### PR TITLE
fix: use undecoded @path, @query and @query-param according to the spec

### PR DESCRIPTION
**Pull request by AlexanderTar**
https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-19#name-path
https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-19#name-query

### CODE REVIEW

**Review by dhensby — APPROVED** [review]
Good catch, thanks. Can we add test coverage for this?

**Comment by dhensby** [comment]
Also, I think this is a fix, not a feat. Would you mind amending the commit message please?

**Comment by dhensby** [comment]
@AlexanderTar alternatively, if you allow maintainers to make pushes to your PR branch, I can make the amendments to the PR

File: src/httpbis/index.ts [hunk 1]
```diff
@@ -112,11 +114,13 @@ export function deriveComponent(component: string, params: Map<string, string |
             if (!params.has('name')) {
                 throw new Error('@query-param must have a named parameter');
             }
-            const name = (params.get('name') as BareItem).toString();
+            const name = decodeURIComponent((params.get('name') as BareItem).toString());
+            console.log(name);
+            console.log(searchParams);
```
Comment (dhensby) [inline src/httpbis/index.ts:119]: Please can you remove the console calls 🙏

File: test/httpbis/httpbis.spec.ts [hunk 2]
```diff
@@ -268,6 +286,31 @@ describe('httpbis', () => {
...
+                    ...req,
+                    url: new URL('https://example.com/parameters?var=this%20is%20a%20big%0Amultiline%20value&bar=with+plus+whitespace&fa%C3%A7ade%22%3A%20=something'),
+                })).to.deep.equal([
+                    'this%20is%20a%20big%0Amultiline%20value',
+                ]);
+                expect(httpbis.deriveComponent('@query-param', new Map([['name', 'bar']]), {
+                    ...req,
+                    url: new URL('https://example.com/parameters?var=this%20is%20a%20big%0Amultiline%20value&bar=with+plus+whitespace&fa%C3%A7ade%22%3A%20=something'),
+                })).to.deep.equal([
+                    'with%20plus%20whitespace',
```
Comment (dhensby) [inline test/httpbis/httpbis.spec.ts:306]: OK - so this is interesting, it's actually not returning a data that was originally in the query string, which means there will be problems when computing signatures.

It seems that once the query string is parsed and stored in the `searchParams` prop, it is fully URL decoded and there's no way to get your hands on the raw value(s).

The only way to get the raw value is using `url.search` but that will need to be split and parsed into parts, which is going to be a bit of a pain.

**Review by dhensby — CHANGES_REQUESTED** [review]
Having to re-encode the query parameters is leading to inconsistencies and edge-case issues, really we need a way to get our hands on the raw value of the query string parameter and, I'm afraid, that looks like it'll mean manual parsing of the `url.search` prop.

File: src/httpbis/index.ts (hunk 1 above)
Comment (AlexanderTar) [reply src/httpbis/index.ts:119]: Oops, artefacts of debugging

**Comment by AlexanderTar** [comment]
> Having to re-encode the query parameters is leading to inconsistencies and edge-case issues, really we need a way to get our hands on the raw value of the query string parameter and, I'm afraid, that looks like it'll mean manual parsing of the `url.search` prop.

According to the spec, this is the process for deriving @query-param component:

The value of the name parameter and the component value of a single named parameter are calculated by the following process:

1. Parse the nameString or valueString of the named query parameter defined by the "application/x-www-form-urlencoded parsing" section of [[HTMLURL](https://url.spec.whatwg.org/#application/x-www-form-urlencoded)], which is the value after percent-encoded octets are decoded.
2. Encode the nameString or valueString using the "percent-encode after encoding" process defined by the "application/x-www-form-urlencoded serializing" section of [[HTMLURL](https://url.spec.whatwg.org/#application/x-www-form-urlencoded)], which results in an [[ASCII](https://www.rfc-editor.org/rfc/rfc20)] string.
3. Output the ASCII string

My research showed that `searchParams` contains a map of URI decoded keys and values, so if `name` was provided with escaped characters, we have to decode it first before looking up. Similarly, once we found a respective value, it has to be encoded. I would imagine all edge cases should be covered by URL parsing code and as long as they are, we're safe to do encode-decode exercises

File: test/httpbis/httpbis.spec.ts (hunk 2 above)
Comment (AlexanderTar) [reply test/httpbis/httpbis.spec.ts:306]: I took examples from the spec and I can't see any problems with the logic in my code: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures#name-query-parameters

Comment (AlexanderTar) [reply test/httpbis/httpbis.spec.ts:306]: Please, note, I'm using `decodeURIComponent` instead of `decodeURI` which apparently does the trick with different representation as part of full URL or simple string

Comment (AlexanderTar) [reply test/httpbis/httpbis.spec.ts:306]: https://stackoverflow.com/a/747700

Comment (dhensby) [reply test/httpbis/httpbis.spec.ts:355]: Ah! Great, I hadn't realised the spec specifically wanted this decode/recode logic. I think the spec here is likely to cause some clients to struggle to be compliant, but that's not our problem!

**Review by dhensby — APPROVED** [review]

**Comment by dhensby** [comment]
Great - thanks for the clarification and the work put into these PRs! 🎉
